cip/CIP2015-09-16-public-type-system-type-annotation.adoc

```diff
@@ -0,0 +1,259 @@
+Stefan Plantikow <stefan.plantikow@neotechnology.com>
+
+toc::[]
+
+== Abstract
+
+This CIP specifies the public part of the type system of Cypher and introduces literal syntax for writing down types as well as new syntax for type annotations. The proposed type annotation syntax does not change the Cypher grammar right now but rather is intended as a template to be used by future CIPs. This CIP also informatively discusses future syntax that might use the introduced literal type syntax.
+
+== Introduction
+=== Background and Motivation
+In order to support currently ongoing work for allowing users to define _procedures_ and _functions_ that interact with Cypher, it has become necessary to specify types explicitly, especially as part of procedure and function signatures. It is thus necessary to at least introduce syntax for all valid, supported public types as well as syntax for type ascription.
+
+=== Overview
+For the purposes of this CIP, a type system is a system of classifying expressions according to what kinds of values they may evaluate to during the execution of a Cypher query. It does *not *include a specification of a type inferencing and checking algorithm for Cypher queries.
+
+Cypher today has a rich type system that unifies Java implementation types into a coherent type hierarchy and an accompanying type checker that supports rich types with generics and type unions. So far it has not been possible to explicitly specify the type of an expression in a query.
+
+This CIP proposes to introduce syntax for type literals and for type annotations.
+
+== 3. Proposal
+=== Types and Type Literal Syntax
+Below we list all currently supported types together with the proposed type literal syntax (in *[]*)
+
+==== Types
+Cypher currently supports values with the following types:
+
+ * `ANY`
+ ** Supertype to all other types
+ * `BOOLEAN`
+ ** true and false
+ * `INTEGER`
+ ** Exact numbers without decimals, i.e. -3, 0, 4
+ * `FLOAT`
+ ** IEEE-754 64-bit floating point numbers
+ * `IDENTITY`
+ ** Identities are considered part of the current type system (since the +++<u>Identity type CIP</u>+++ has been accepted), even though it has not been implemented at the time of writing this CIP
+ * `MAP`
+ ** Maps to values of any type, i.e. { name: "Svensson" }. Note that keys are identifiers, not strings.
+ ** This is also the supertype for `NODE` and `RELATIONSHIP`
+ * `LIST<T>`
+ ** Lists (ordered sequences with random access) of a given type T
+ ** `LIST<T1>`` is a subtype of `LIST<T2>`` if `T1` is a subtype of `T2`. This works since values in Cypher are immutable. Adding an element of type `T` to a `LIST<S>`` would produce a new list of type `LIST<R>``, where `R` is the nearest common supertype of `T` and `S`
+ * `NODE`
+ ** Nodes are maps that have labels and relationships
+ * `STRING`
+ ** Unicode Strings, i.e. `"Cypher"`, and `‘text’`
+ * `RELATIONSHIP`
+ ** Relationships are maps that have a type and end points
```

Comment:
I have the same irk here as I do with the definition of node. Actually here the definition of what a relationship is feels even stranger than what the prior definition of node did.